include/cuda/memory_resource

```diff
@@ -0,0 +1,309 @@
+#if _LIBCUDACXX_STD_VER < 17
+_LIBCUDACXX_CONCEPT resource_with =
+    resource<_Resource> &&_CUDA_VSTD::conjunction<_CUDA_VSTD::bool_constant<
+        has_property<_Resource, _Properties>>...>::value;
+#else
+_LIBCUDACXX_CONCEPT resource_with = resource<_Resource> //
+                                    && (has_property<_Resource, _Properties> &&
+                                        ...);
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+// properties
+template <class...> using _Void_t = void;
+
+template <class _Property, typename = void>
+_LIBCUDACXX_INLINE_VAR constexpr bool property_with_value_v = false;
+template <class _Property>
+_LIBCUDACXX_INLINE_VAR constexpr bool
+    property_with_value_v<_Property, _Void_t<typename _Property::value_type>> =
+        true;
+
+template <class _Property>
+using property_value_t = _CUDA_VSTD::_If<property_with_value_v<_Property>,
+                                         typename _Property::value_type, void>;
+
+/// class any_resource_ref
+// clang-format off
+struct _Resource_vtable_builder
+{
+    template <class _Resource, class _Property>
+    inline __host__ __device__
+    static typename _Property::value_type _Get_property(void *__res) noexcept {
+        return get_property(*static_cast<const _Resource *>(__res), _Property{});
+    }
+
+    template <class _Resource>
+    inline __host__ __device__ 
+    static void* _Alloc(void *__object, size_t __bytes, size_t __alignment) {
+        return static_cast<_Resource *>(__object)->allocate(__bytes, __alignment);
+    }
+
+    template <class _Resource>
+    inline __host__ __device__ 
+    static void _Dealloc(void *__object, void *__ptr, size_t __bytes, size_t __alignment) {
+        return static_cast<_Resource *>(__object)->deallocate(__ptr, __bytes, __alignment);
+    }
+
+    template <class _Resource>
+    inline __host__ __device__ 
+    static bool _Equal(void *__left, void *__right) {
+        return *static_cast<_Resource *>(__left) == *static_cast<_Resource *>(__right);
+    }
+};
+// clang-format on
+
+template <class _Property> //
+struct _Property_vtable {
+  using _PropertyFn = typename _Property::value_type (*)(void *);
+  _PropertyFn __property_fn = nullptr;
+
+  // clang-format off
+  inline __host__ __device__ 
+  constexpr _Property_vtable(_PropertyFn __property_fn_) noexcept
+      : __property_fn(__property_fn_) {}
+  // clang-format on
+};
+
+template <class... _Properties>
+struct _Resource_vtable : public _Property_vtable<_Properties>... {
+  using _AllocFn = void *(*)(void *, size_t, size_t);
+  using _DeallocFn = void (*)(void *, void *, size_t, size_t);
+  using _EqualFn = bool (*)(void *, void *);
+
+  _AllocFn __alloc_fn = nullptr;
+  _DeallocFn __dealloc_fn = nullptr;
+  _EqualFn __equal_fn = nullptr;
+
+  template <class... _PropertyFns>
+  inline __host__ __device__ //
+      constexpr _Resource_vtable(_AllocFn __alloc_fn_, _DeallocFn __dealloc_fn_,
```

Comment:
Generally, I would have expected to be able to initialize this with list-initialization as it should be an aggregate. Is this some device side limitation?

Reply:
Why future me, thanks for reminding me that inheriting from a base classes makes this not an aggregate...

Reply:
In C++17 you can if the base is public and also an aggregate.

Reply:
I feel like "in C++17 you could ...." will be a reoccurring theme in the near future 😹